---
title: 'BMI-brain geneset tests'
author: "Jon Thompson, Pers lab"
date: "`r Sys.time()`"
output:
  html_notebook:
    df_print: paged
    number_sections: yes
    toc: yes
    toc_depth: 3
    toc_float: yes
---
# setup


## source
```{r}
source(file="/projects/jonatan/tools/functions-src/utility_functions.R")
source(file="/projects/jonatan/tools/functions-src/functions_sc.R")
```

## packages

```{r}

ipak(c("Matrix","dplyr", "parallel", "reshape", "reshape2", "ggplot2", "plotly"))

```

## options

```{r}
options(stringsAsFactors = F, use= "pairwise.complete.obs", warn=1)
```

## constants

```{r}
ncelltypes =11
p.val.threshold = 5e-2

dirOut <- "/projects/jonatan/applied/18-mousebrain_7/"
# Set paths
dirTables = paste0(dirOut, "tables/")
dirLog <- paste0(dirOut, "log/")
dirRObjects <- paste0(dirOut, "RObjects/")
dirPlots <- paste0(dirOut, "plots/")
dirScratch = "/scratch/jonatan/"

# Make directories
for (direc in list(dirLog, dirRObjects, dirPlots)) {
  if (!file.exists(direc)) {
    dir.create(direc)
  }
}

flag_date = substr(gsub("-","",as.character(Sys.Date())),3,1000)
prefixOut <- paste0("mb_7.2_", flag_date)
```

## load data

## load files

```{r}
df_NWA<-load_obj(paste0(dirTables, "Neurons_sub_ClusterName_7.2_run1_cell_cluster_module_genes.csv.gz"))
```

<!-- LDSCORE gene module results -->

<!-- ```{r} -->
<!-- df_celltyperesults <- read.delim("/raid5/projects/timshel/sc-genetics/sc-genetics/out/out.ldsc/wgcna.mousebrain-190111.fdr_sign_celltypes.continuous__BMI_Yengo2018.cell_type_results.txt", sep = "\t", header=T) -->
<!-- head(df_celltyperesults) -->
<!-- ``` -->

load and format genesets as a single list
(run once)
```{r}
if (FALSE) {
  vec_path_geneset <-
    c("/raid5/projects/timshel/sc-genetics/sc-genetics/data/genes_obesity/turcot2018_s21_mendelian_obesity_genes.mapped.txt",
      "/raid5/projects/timshel/sc-genetics/sc-genetics/data/genes_obesity/turcot2018_table1_rare_variants.mapped.txt",
      "/raid5/projects/timshel/sc-genetics/sc-genetics/data/genes_obesity/yazdi2015_table1_mouse_obesity_genes.mapped.txt",
      "/raid5/projects/timshel/sc-genetics/sc-genetics/data/genes_predefined_gene_sets/human_psd_genes.pers_hmg2016_table_s12.txt",
      "/raid5/projects/timshel/sc-genetics/sc-genetics/data/genes_predefined_gene_sets/mouse_psd_genes.pers_hmg2016_table_s13.txt")
  
  list_df_geneset <- lapply(vec_path_geneset, load_obj)
  str(list_df_geneset)
  list_geneset <- lapply(list_df_geneset, function(df){df[[grep("^ensembl_.*id$",colnames(df))]]})
  names(list_geneset) <- c("turcot2018_s21_mendelian_obesity_genes.mapped", "turcot2018_table1_rare_variants.mapped", "yazdi2015_table1_mouse_obesity_genes.mapped", "human_psd_genes.pers_hmg2016_table_s12", "mouse_psd_genes.pers_hmg2016_table_s13")
  saveRDS(list_geneset, file = "/projects/jonatan/genesets/bmi_brain_hsapiens.RDS")
}
```

Make shorter vectors of celltype ES values to use in Fisher's exact test
(run once)
```{r}
if (FALSE){
  df_SEM <- load_obj("/projects/timshel/sc-genetics/sc-genetics/data/genes_cell_type_specific/mousebrain_all.mean.csv.gz")
  regExCelltypes <- "^MEGLU10$|^MEGLU11$|^TEGLU17$|^DEINH3$|^MEGLU1$|^MEINH2$|^TEGLU23$|^DEGLU5$|^TEGLU4|^DEGLU4$|^TEINH12$"
  vec_idxColCelltype <- grep("^MEGLU10$|^MEGLU11$|^TEGLU17$|^DEINH3$|^MEGLU1$|^MEINH2$|^TEGLU23$|^DEGLU5$|^TEGLU4|^DEGLU4$|^TEINH12$", colnames(df_SEM))
  
  cutoff <- 0.9
  
  list_vec <- sapply(df_SEM[,vec_idxColCelltype], function(vec_col) {
    names(vec_col) <- df_SEM[,1]
    vec_col <- vec_col[vec_col>=cutoff] 
    sort(vec_col, decreasing=T)
  })
  saveRDS(list_vec, file = paste0("/projects/jonatan/mousebrain_7/RObjects/list_mb_SEM_signif_", cutoff, "cutoff.RDS.gz"), compress="gzip")
}
```


## plotting BMI enrichment

Load BMI enrichment matrix

```{r}
#mat_SEMtop_BMI_fisher_pval <- read.delim("/projects/jonatan/mousebrain_7/tables/mb_GSA_SEMtop0.75_BMI_1_mat_fisherPval_genesetTests.txt", sep = "\t", header=T)
vec_path_mat <- dir(path=dirTables, pattern = "mb_ALL_GSA_.*1.3", full.names = T)
vec_path_mat <- vec_path_mat[!grepl("meta",vec_path_mat)]
```

```{r}
path_mat_SEMall_BMI_wilcoxon_pval <- grep(pattern = "mb_priorwilcoxonPval",x=vec_path_mat,value=T)
mat_SEMall_BMI_wilcoxon_pval <- load_obj(path_mat_SEMall_BMI_wilcoxon_pval)
```

```{r}
mat_testUse <- mat_SEMall_BMI_wilcoxon_pval
n_tests = dim(mat_testUse)[2]
```

```{r}
mat_testUse %>% log10 %>% '-'(.) -> mat_testUse_log
mat_testUse_log$geneset <- rownames(mat_testUse_log)
df_long <- melt(data = mat_testUse_log, id.vars = "geneset")
#celltype <- rep(names(list_vec[[1]]), rep(length(modules),11))
colnames(df_long) <- c("test", "celltype", "minlog10pval")

(p.val.threshold / n_tests) %>% log10(.) %>% '-'(.) -> cutoff

# Grouped

ggplot(df_long, aes(fill=test, y=minlog10pval, x=celltype)) +
  geom_hline(yintercept=cutoff, linetype="dashed", color = "red") +
  geom_bar(position="dodge", stat="identity") + theme(axis.text.x = element_text(angle = 90, hjust = 1))
saveMeta(savefnc = ggsave, filename = paste0(dirPlots, prefixOut, "_SEM_BMI_enrich.pdf"), width=40, height=10)
```

### BMI enrichment test on ALL mb celltypes: wilcoxon and GSEA

```{r}
mat_wilcoxon_pval <- load_obj("/projects/jonatan/applied/18-mousebrain_7/tables/mb_ALL_GSA_SEMall_BMI_1.3_analytical_mat_wilcoxonPval_genesetTests.txt")
mat_GSEA_pval <- load_obj("/projects/jonatan/applied/18-mousebrain_7/tables/mb_ALL_GSA_SEMall_BMI_1.4_1000_mat_GSEAPval_genesetTests.txt")
mat_t.test_pval <- load_obj("/projects/jonatan/applied/18-mousebrain_7/tables/mb_ALL_GSA_SEMall_BMI_1.4_1000_mat_t.testPval_genesetTests.txt")
mat_GSEA_sscore <- load_obj("/projects/jonatan/applied/18-mousebrain_7/tables/mb_ALL_GSA_SEMall_BMI_1.4_1000_mat_GSEAsscore_genesetTests.txt")
```

```{r}
mat_1 <- mat_GSEA_pval
mat_2 <- mat_t.test_pval
mat_3 <- mat_wilcoxon_pval
mat_4 <- mat_GSEA_sscore
```

```{r}
all.equal(colnames(mat_1), colnames(mat_2))
all.equal(rownames(mat_1), rownames(mat_2))
all.equal(colnames(mat_1), colnames(mat_3))
all.equal(rownames(mat_1), rownames(mat_3))
all.equal(colnames(mat_1), colnames(mat_4))
all.equal(rownames(mat_1), rownames(mat_4))

```

### transform pvals to -log10

```{r}
mat_1_pvalNegLog10 <- -log10(mat_1)
mat_2_pvalNegLog10 <- -log10(mat_2)
mat_3_pvalNegLog10 <- -log10(mat_3)

```

### how correlated are enrichment metrics?

```{r}
mat_cor_mat_2_mat_GSEAsscore <- cor(t(mat_2_pvalNegLog10), t(mat_GSEA_sscore))
mat_cor_mat_1_mat_2 <- cor(t(mat_1_pvalNegLog10), t(mat_2_pvalNegLog10))
mat_cor_mat_2_mat_3 <- cor(t(mat_2_pvalNegLog10), t(mat_3_pvalNegLog10))
#mat_cor_wilcoxon_GSEAPvalNegLog10 <- cor(t(mat_wilcoxon_pvalNegLog10), t(mat_GSEA_pvalNegLog10))
```


### 

The correlations between the wilcoxon and the GSEA tests are surprisingly low:

```{r}
diag(mat_cor_wilcoxon_GSEAPvalNegLog10)
diag(mat_cor_wilcoxon_GSEAsscore)
```

Compute correlation between all enrichment measures on yazdi 2015

```{r}
vec_path <- dir(path = dirTables, pattern = "yazdi_2015.*", full.names=T)
list_mat <- lapply(vec_path, load_obj)
vec_path %>% gsub(dirTables,"", .)  %>% gsub("/mb_GSA_SEMall_BMI_yazdi_2015_1?_?|_genesetTests.txt","",.) ->names(list_mat)
```

```{r}

list_mat <- lapply(list_mat, function(eachmat) {
  if (nrow(eachmat)<ncol(eachmat)) {
    eachmat <- t(eachmat)
  }
  return(eachmat)
})

mat_allTestsPval <-Reduce(f=cbind, x = list_mat[grep("Pval",names(list_mat))])
rownames(mat_allTestsPval) <- rownames(list_mat[["mat_wilcoxonPval"]])
colnames(mat_allTestsPval) <- names(list_mat)[grep("Pval",names(list_mat))]

```

We see that most enrichment tests are more pairwise correlated than wilcoxon & GSEA 

```{r}
mat_cor_allTestsYazdi <- cor(mat_allTestsPval) 
mat_cor_allTestsYazdi
```

## enrichment heatmaps

### all celltypes

```{r}
list_mat1 <- list("GSEA"=mat_GSEA_pvalNegLog10, 
                  "wilcoxon"=mat_wilcoxon_pvalNegLog10) 

for (name in names(list_mat1)) {
  mat <- list_mat1[[name]]
  df <- data.frame(mat)
  df <- cbind("geneset" = rownames(df), df)
  df.m <- melt(t(mat), id.vars=geneset)
  colnames(df.m) <- c("celltype","geneset","minLog10pval")
  
 p <- ggplot(df.m, aes(geneset, celltype)) + 
     geom_tile(aes(fill = minLog10pval),colour = "white") + 
     scale_fill_gradientn(colours=c("cadetblue1","red"))+
     theme(axis.text.x = element_text(angle=90))
 
 #p <- ggplotly(p)
 
  ggsave(plot=p, filename = paste0(dirPlots, prefixOut, "_",name, "_ALL_SEM_heatmap.pdf"), width=3.5, height=32)
 
}
```

### SEM enriched celltypes

```{r}
list_mat1 <- list("GSEA"=mat_GSEA_pvalNegLog10, 
                  "wilcoxon"=mat_wilcoxon_pvalNegLog10) 

for (name in names(list_mat1)) {
  mat <- list_mat1[[name]]
  df <- data.frame(mat)
  df <- cbind("geneset" = rownames(df), df)
  df.m <- melt(t(mat), id.vars=geneset)
  colnames(df.m) <- c("celltype","geneset","minLog10pval")
  
 p <- ggplot(df.m, aes(geneset, celltype)) + 
     geom_tile(aes(fill = minLog10pval),colour = "white") + 
     scale_fill_gradientn(colours=c("cadetblue1","red"))+
     theme(axis.text.x = element_text(angle=90))
 
 #p <- ggplotly(p)
 
  ggsave(plot=p, filename = paste0(dirPlots, prefixOut, "_",name, "_ALL_SEM_heatmap.pdf"), width=3.5, height=32)
 
}
```

enrichment results: make boxplots 
TODO

```{r}
df_mapping = load_obj("/projects/timshel/sc-genetics/sc-genetics/data/gene_annotations/gene_annotation.hsapiens_mmusculus_unique_orthologs.GRCh37.ens_v91.txt.gz")
df_NWA[["ensembl_HS"]] <- gene_map(dataIn = df_NWA[["ensembl"]], df_mapping = df_mapping, from="mmusculus_homolog_ensembl_gene", to="ensembl_gene_id", na.rm = F, replace=F)
```

```{r}
modules = c("lavenderblush", "lightpink3")
df_data <- df_SEM[,c(1, grep("^MEGLU10$|^MEGLU11$|^TEGLU17$|^DEINH3$|^MEGLU1$|^MEINH2$|^TEGLU23$|^DEGLU5$|^TEGLU4|^DEGLU4$|^TEINH12$", colnames(df_SEM)))]
df_data[["lavenderblush"]] = df_SEM[["gene"]]  %in% df_NWA[["ensembl_HS"]][df_NWA[["module"]]=="lavenderblush"]
df_data[["lightpink3"]] = df_SEM[["gene"]]  %in% df_NWA[["ensembl_HS"]][df_NWA[["module"]]=="lightpink3"]
df_data_long <- reshape::melt(df_data, id.vars=c("gene","lavenderblush", "lightpink3"))
colnames(df_data_long)[4] <- "celltype"
df_data_long
```

```{r, fig.width = 15, fig.height=10}
ggplot(data = df_data_long, aes(lavenderblush, fill=celltype, value)) +
geom_boxplot(stat = "boxplot") +
facet_wrap(facets = vars(celltype), nrow=1, scales="fixed") +
theme_minimal()
```

