---
title: 'BMI-brain plotting'
author: "Jon Thompson, Pers lab"
date: "`r Sys.time()`"
output:
  html_notebook:
    df_print: paged
    number_sections: yes
    toc: yes
    toc_depth: 3
    toc_float: yes
  html_document:
    df_print: paged
    number_sections: yes
    toc: yes
    toc_depth: 3
    toc_float: yes
---
# setup

```{r}
# for network plots, load WGCNA session image. Do this before sourcing functions since the session image functions will be dated
load("/projects/jonatan/applied/18-mousebrain_7/RObjects/Neurons_sub_ClusterName_7.2_run1_final_session_image.RData.gz")
```

## source
```{r}
source(file="/projects/jonatan/tools/functions-src/utility_functions.R")
source(file="/projects/jonatan/tools/functions-src/functions_sc.R")
```

## packages

```{r}

ipak(c("Matrix","dplyr", "parallel", "reshape", "reshape2", "ggplot2", "plotly", "ComplexHeatmap"))
#ipak(c("ComplexHeatmap", "igraph", "WGCNA", "RColorBrewer"))
#ipak(c("corrr", "ggraph", "here", "tidygraph", "tidyverse"))

```

## options

```{r}
options(stringsAsFactors = F, use= "pairwise.complete.obs", warn=1)
```

## constants

```{r}

ncelltypes =11
p.val.threshold = 5e-2

dirOut <- "/projects/jonatan/applied/18-mousebrain_7/"
# Set paths
dirTables = paste0(dirOut, "tables/")
dirLog <- paste0(dirOut, "log/")
dirRObjects <- paste0(dirOut, "RObjects/")
dirPlots <- paste0(dirOut, "plots/")
dirScratch = "/scratch/jonatan/"

# Make directories
for (direc in list(dirLog, dirRObjects, dirPlots)) {
  if (!file.exists(direc)) {
    dir.create(direc)
  }
}

flag_date = substr(gsub("-","",as.character(Sys.Date())),3,1000)
prefixOut <- paste0("mb_7.2_", flag_date)
```

## load data

<!-- LDSCORE gene module results -->

<!-- ```{r} -->
<!-- df_celltyperesults <- read.delim("/raid5/projects/timshel/sc-genetics/sc-genetics/out/out.ldsc/wgcna.mousebrain-190111.fdr_sign_celltypes.continuous__BMI_Yengo2018.cell_type_results.txt", sep = "\t", header=T) -->
<!-- head(df_celltyperesults) -->
<!-- ``` -->

# not run
# load and format gene lists

```{r}
if (FALSE) {
  vec_path_geneset <-
    c("/raid5/projects/timshel/sc-genetics/sc-genetics/data/genes_obesity/turcot2018_s21_mendelian_obesity_genes.mapped.txt",
      "/raid5/projects/timshel/sc-genetics/sc-genetics/data/genes_obesity/turcot2018_table1_rare_variants.mapped.txt",
      "/raid5/projects/timshel/sc-genetics/sc-genetics/data/genes_obesity/yazdi2015_table1_mouse_obesity_genes.mapped.txt",
      "/raid5/projects/timshel/sc-genetics/sc-genetics/data/genes_predefined_gene_sets/human_psd_genes.pers_hmg2016_table_s12.txt",
      "/raid5/projects/timshel/sc-genetics/sc-genetics/data/genes_predefined_gene_sets/mouse_psd_genes.pers_hmg2016_table_s13.txt")
  
  list_df_geneset <- lapply(vec_path_geneset, load_obj)
  str(list_df_geneset)
  list_geneset <- lapply(list_df_geneset, function(df){df[[grep("^ensembl_.*id$",colnames(df))]]})
  names(list_geneset) <- c("turcot2018_s21_mendelian_obesity_genes.mapped", "turcot2018_table1_rare_variants.mapped", "yazdi2015_table1_mouse_obesity_genes.mapped", "human_psd_genes.pers_hmg2016_table_s12", "mouse_psd_genes.pers_hmg2016_table_s13")
  saveRDS(list_geneset, file = "/projects/jonatan/genesets/bmi_brain_hsapiens.RDS")
}
```

### not run 
```{r}
if (FALSE){
  df_SEM <- load_obj("/projects/timshel/sc-genetics/sc-genetics/data/genes_cell_type_specific/mousebrain_all.mean.csv.gz")
  regExCelltypes <- "^MEGLU10$|^MEGLU11$|^TEGLU17$|^DEINH3$|^MEGLU1$|^MEINH2$|^TEGLU23$|^DEGLU5$|^TEGLU4|^DEGLU4$|^TEINH12$"
  vec_idxColCelltype <- grep("^MEGLU10$|^MEGLU11$|^TEGLU17$|^DEINH3$|^MEGLU1$|^MEINH2$|^TEGLU23$|^DEGLU5$|^TEGLU4|^DEGLU4$|^TEINH12$", colnames(df_SEM))
  
  cutoff <- 0.9
  
  list_vec <- sapply(df_SEM[,vec_idxColCelltype], function(vec_col) {
    names(vec_col) <- df_SEM[,1]
    vec_col <- vec_col[vec_col>=cutoff] 
    sort(vec_col, decreasing=T)
  })
  saveRDS(list_vec, file = paste0("/projects/jonatan/mousebrain_7/RObjects/list_mb_SEM_signif_", cutoff, "cutoff.RDS.gz"), compress="gzip")
}
```


```{r}
## SEM versus WGCNA
### SEM all versus WGCNA top
#### t.test
mat_SEMall_WGCNAtop_t.test_pval <- read.delim("/projects/jonatan/mousebrain_7/tables/mb_GSA_SEMall_WGCNAtop_1_mat_t.test_pval_genesetTests.txt", sep = "\t", header=T)
#### Wilcoxon
mat_SEMall_WGCNAtop_wilcoxon_pval <- read.delim("/projects/jonatan/mousebrain_7/tables/mb_GSA_SEMall_WGCNAtop_1_mat_wilcoxonPval_genesetTests.txt", sep = "\t", header=T)
#### GSEA
mat_SEMall_WGCNAtop_GSEA_pval <- read.delim("/projects/jonatan/mousebrain_7/tables/mb_GSA_SEMall_WGCNAtop_1_mat_GSEAPval_genesetTests.txt",sep = "\t", header=T)
#mat_SEMall_WGCNAtop_GSEA_sscore  <- read.delim("/projects/jonatan/mousebrain_7/tables/mb_GSA_SEMall_WGCNAtop_1_mat_GSEAsscore_genesetTests.txt")
#### dot.product
mat_SEMall_WGCNAtop_dot.product_pval <- read.delim("/projects/jonatan/mousebrain_7/tables/mb_GSA_SEMall_WGCNAtop_1_mat_dot.product.testPval_genesetTests.txt")
### SEM top versus WGCNA top
#### Fisher
mat_SEMtop_WGCNAtop_fisher_pval<- read.delim("/projects/jonatan/mousebrain_7/tables/mb_GSA_SEMtop150_WGCNAtop_1_mat_fisherPval_genesetTests.txt", sep = "\t", header=T)

## WGCNA versus SEM
### WGCNA all versus SEM top
#### t.test
mat_WGCNAall_SEMtop_t.test_pval <- read.delim("/projects/jonatan/mousebrain_7/tables/mb_GSA_WGCNAall_SEMtop150_1_mat_t.testPval_genesetTests.txt", sep = "\t", header=T)
### wilcoxon
mat_WGCNAall_SEMtop_wilcoxon_pval <- read.delim("/projects/jonatan/mousebrain_7/tables/mb_GSA_WGCNAall_SEMtop150_1_mat_wilcoxonPval_genesetTests.txt", sep = "\t", header=T)
### GSEA
mat_WGCNAall_SEMtop_GSEA_pval <- read.delim("/projects/jonatan/mousebrain_7/tables/mb_GSA_WGCNAall_SEMtop150_1_mat_GSEAPval_genesetTests.txt", sep = "\t", header=T)
#mat_WGCNAall_SEMtop_GSEA_sscore <- read.delim("/projects/jonatan/mousebrain_7/tables/mb_GSA_WGCNAall_SEMtop150_1_mat_GSEAsscore_genesetTests.txt", sep = "\t", header=T)
### dot.product
mat_WGCNAall_SEMtop_dot.product_pval <- read.delim("/projects/jonatan/mousebrain_7/tables/mb_GSA_WGCNAall_SEMtop150_1_mat_dot.product.testPval_genesetTests.txt", sep = "\t", header=T)

### WGCNA all versus SEM all
#### dot.product
mat_SEMall_WGCNAall_dot.product_pval <- read.delim("/projects/jonatan/mousebrain_7/tables/mb_GSA_SEMall_WGCNAall_1_mat_dot.product.testPval_genesetTests.txt", sep = "\t", header=T)

```

Compare different SEM cutoffs for Fisher's exact test

```{r}
#mat_SEMtop_WGCNAtop_fisher_pval150 <- read.delim("/projects/jonatan/mousebrain_7/tables/mb_GSA_SEMtop150_WGCNAtop_1_mat_fisherPval_genesetTests.txt", sep = "\t", header=T)
mat_SEMtop_WGCNAtop_fisher_pvaltop150 <- read.delim("/projects/jonatan/mousebrain_7/tables/mb_GSA_SEMtop150_WGCNAtop_1_mat_fisherPval_genesetTests.txt", sep = "\t", header=T)
mat_SEMtop_WGCNAtop_fisher_pval0.5 <- read.delim("/projects/jonatan/mousebrain_7/tables/mb_GSA_SEMtop0.5_WGCNAtop_1_mat_fisherPval_genesetTests.txt", sep = "\t", header=T)
mat_SEMtop_WGCNAtop_fisher_pval0.75<- read.delim("/projects/jonatan/mousebrain_7/tables/mb_GSA_SEMtop0.75_WGCNAtop_1_mat_fisherPval_genesetTests.txt", sep = "\t", header=T)
mat_SEMtop_WGCNAtop_fisher_pval0.9 <- read.delim("/projects/jonatan/mousebrain_7/tables/mb_GSA_SEMtop0.9_WGCNAtop_1_mat_fisherPval_genesetTests.txt", sep = "\t", header=T)
```

```{r}
df_NWA<-load_obj(paste0(dirTables, "Neurons_sub_ClusterName_7.2_run1_cell_cluster_module_genes.csv.gz"))
```

### Get celltype scores of modules on different tests

### plotting

```{r}

modules <- c("lavenderblush", "lightpink3")
list_mat <- ls(pattern="mat_.*_pval") %>% lapply(., function(matName) eval(parse(text=matName)))
names(list_mat) <-  ls(pattern="mat_.*_pval")
n_tests = length(modules) * ncelltypes

```

#### Compare different enrichment metrics

```{r}

for (module in modules){ 
  
  list_vec <- lapply(list_mat, function(mat_tmp) {
    mat_tmp <- as.matrix(mat_tmp)
    if (nrow(mat_tmp)<ncol(mat_tmp)) mat_tmp <- t(mat_tmp)
    #mat_tmp[module,] %>% p.adjust(p = ., method = "fdr", n = n_tests) %>% log10 %>% '-'(.) -> vec_tmp
    mat_tmp[module,] %>% log10 %>% '-'(.) -> vec_tmp 
    
    #mat_tmp[module,]  -> vec_tmp
    names(vec_tmp) <- colnames(mat_tmp)
    vec_tmp
  })
  #ls(pattern="mat_.*_pval") %>% gsub("mat_", "", .) %>% gsub("_pval", "qval", .) -> names(list_vec) 
  ls(pattern="mat_.*_pval") %>% gsub("mat_", "", .) -> names(list_vec) 
  
  # create a dataset
  
  celltype <- rep(names(list_vec[[1]]), rep(length(list_vec),11))
  
  test <- rep(names(list_vec),11)
  
  list_vec %>% as.data.frame %>% as.matrix %>% t %>% as.numeric -> vec_vals
  # matrix(data = unlist(list_vec), ncol=9) %>% t-> mat_vals
  # vec_vals <- as.numeric(mat_vals)
  df_data <- data.frame(celltype=celltype, test=test, value=vec_vals)
  #df_data
  
  (p.val.threshold / n_tests) %>% log10(.) %>% '-'(.) -> cutoff
  # Grouped
  ggplot(df_data, aes(fill=test, y=value, x=celltype)) +
    geom_bar(position="dodge", stat="identity") + theme(axis.text.x = element_text(angle = 90, hjust = 1)) + 
    geom_hline(yintercept=cutoff, linetype="dashed", color = "red") 
  ggsave(filename = paste0(dirPlots, prefixOut, "_",module ,"_modCelltypeEnrich_compare.pdf"), width=15, height=10)

}
```

#### Plot module celltype specificity barplot 
using selected enrichment metric

```{r}
mat_testUse <- mat_WGCNAall_SEMtop_wilcoxon_pval
name <- "wilcoxon"
#mat_testUse <- mat_SEMtop_WGCNAtop_fisher_pval
#mat_testUse <- mat_SEMall_WGCNAtop_wilcoxon_pval
if (nrow(mat_testUse)<ncol(mat_testUse)) mat_testUse <- t(mat_testUse)
mat_testUse_modules <- mat_testUse[modules,]
mat_testUse_modules %>% as.matrix %>% as.numeric %>% p.adjust(p = ., method = "fdr", n = n_tests) %>% log10 %>% '-'(.) -> vec_vals

celltype <- rep(names(list_vec[[1]]), rep(length(modules),11))
vec_module <- rep(modules, 11)
df_data1 <- data.frame(celltype=celltype, module=vec_module, value=vec_vals)
df_data1

# Grouped
ggplot(df_data1, aes(fill=module, y=value, x=celltype)) +
  geom_bar(position="dodge", stat="identity") + 
  theme(axis.text.x = element_text(angle = 90, hjust = 1))
ggsave(filename = paste0(dirPlots, prefixOut, paste0(modules, collapse=",_"),"_", name, "_celltypeEnrich.pdf"), width=15, height=10)
```

#### celltype BMI enrichment
Load BMI enrichment matrix

```{r}
#mat_SEMtop_BMI_fisher_pval <- read.delim("/projects/jonatan/mousebrain_7/tables/mb_GSA_SEMtop0.75_BMI_1_mat_fisherPval_genesetTests.txt", sep = "\t", header=T)
vec_path_mat <- dir(path=dirTables, pattern = "BMI_1", full.names = T)
```

```{r}
path_mat_SEMall_BMI_wilcoxon_pval <- grep(pattern = "wilcoxonPval",x=vec_path_mat,value=T)
mat_SEMall_BMI_wilcoxon_pval <- load_obj(path_mat_SEMall_BMI_wilcoxon_pval)

```

```{r}
n_tests =ncelltypes

mat_testUse <- mat_SEMtop_BMI_fisher_pval

mat_testUse %>% log10 %>% '-'(.) -> mat_testUse_log
mat_testUse_log$geneset <- rownames(mat_testUse_log)
df_long <- melt(data = mat_testUse_log, id.vars = "geneset")
#celltype <- rep(names(list_vec[[1]]), rep(length(modules),11))
colnames(df_long) <- c("test", "celltype", "minlog10pval")

(p.val.threshold / n_tests) %>% log10(.) %>% '-'(.) -> cutoff

# Grouped
ggplot(df_long, aes(fill=test, y=minlog10pval, x=celltype)) +
  geom_hline(yintercept=cutoff, linetype="dashed", color = "red") +
  geom_bar(position="dodge", stat="identity") + theme(axis.text.x = element_text(angle = 90, hjust = 1))
ggsave(filename = paste0(dirPlots, prefixOut,"SEM_BMI_fisher_enrich.pdf"), width=15, height=10)
```

### BMI enrichment test on ALL mb celltypes: wilcoxon and GSEA

```{r}
mat_wilcoxon_pval <- load_obj(paste0(dirTables, "mb_ALL_GSA_SEMall_BMI_1_mat_wilcoxonPval_genesetTests.txt"))
mat_GSEA_pval <- load_obj(paste0(dirTables,"mb_ALL_GSA_SEMall_BMI_1.1_mat_GSEAPval_genesetTests.txt"))
mat_GSEA_sscore <- load_obj(paste0(dirTables,"mb_ALL_GSA_SEMall_BMI_1.1_mat_GSEAsscore_genesetTests.txt"))
```

### For GSEA, only keep p-values of positive sscores

```{r}
mat_GSEA_pval[mat_GSEA_sscore < 0] <- 1
```

```{r}
all.equal(colnames(mat_wilcoxon_pval), colnames(mat_GSEA_pval))
all.equal(colnames(mat_GSEA_pval), colnames(mat_GSEA_sscore))
all.equal(rownames(mat_wilcoxon_pval), rownames(mat_GSEA_pval))
all.equal(rownames(mat_GSEA_pval), rownames(mat_GSEA_sscore))
```

### transform pvals to -log10

```{r}
mat_wilcoxon_pvalNegLog10 <- -log10(mat_wilcoxon_pval)
mat_GSEA_pvalNegLog10 <- -log10(mat_GSEA_pval)
```

### how correlated are wilcoxon matrix and GSEA pvals?

```{r}
mat_cor_wilcoxon_GSEAsscore <- cor(t(mat_wilcoxon_pvalNegLog10), t(mat_GSEA_sscore))
mat_cor_wilcoxon_GSEAPvalNegLog10 <- cor(t(mat_wilcoxon_pvalNegLog10), t(mat_GSEA_pvalNegLog10))
```

The correlations between the wilcoxon and the GSEA tests are surprisingly low:

```{r}
diag(mat_cor_wilcoxon_GSEAPvalNegLog10)
diag(mat_cor_wilcoxon_GSEAsscore)
```

Compute correlation between all enrichment measures on yazdi 2015

```{r}
vec_path <- dir(path = dirTables, pattern = "yazdi_2015.*", full.names=T)
list_mat <- lapply(vec_path, load_obj)
vec_path %>% gsub(dirTables,"", .)  %>% gsub("/mb_GSA_SEMall_BMI_yazdi_2015_1?_?|_genesetTests.txt","",.) ->names(list_mat)

# set P-values for negative GSEA sscores to 1
list_mat[["mat_GSEAPval"]][list_mat[["mat_GSEAsscore"]]<0] <- 1 

```

```{r}

list_mat <- lapply(list_mat, function(eachmat) {
  if (nrow(eachmat)<ncol(eachmat)) {
    eachmat <- t(eachmat)
  }
  return(eachmat)
})

mat_allTestsPval <-Reduce(f=cbind, x = list_mat[grep("Pval",names(list_mat))])
rownames(mat_allTestsPval) <- rownames(list_mat[["mat_wilcoxonPval"]])
colnames(mat_allTestsPval) <- names(list_mat)[grep("Pval",names(list_mat))]

```

We see that most enrichment tests are more pairwise correlated than wilcoxon & GSEA 

```{r}
mat_cor_allTestsYazdi <- cor(mat_allTestsPval) 
mat_cor_allTestsYazdi
```

## enrichment heatmaps

### all celltypes

```{r}
list_mat1 <- list("GSEA"=mat_GSEA_pvalNegLog10, 
                  "wilcoxon"=mat_wilcoxon_pvalNegLog10) 

for (name in names(list_mat1)) {
  mat <- list_mat1[[name]]
  df <- data.frame(mat)
  df <- cbind("geneset" = rownames(df), df)
  df.m <- melt(t(mat), id.vars=geneset)
  colnames(df.m) <- c("celltype","geneset","minLog10pval")
  
 p <- ggplot(df.m, aes(geneset, celltype)) + 
     geom_tile(aes(fill = minLog10pval),colour = "white") + 
     scale_fill_gradientn(colours=c("cadetblue1","red"))+
     theme(axis.text.x = element_text(angle=90))
 
 #p <- ggplotly(p)
 
  ggsave(plot=p, filename = paste0(dirPlots, prefixOut, "_",name, "_ALL_SEM_heatmap.pdf"), width=3.5, height=32)
 
}
```

### SEM enriched celltypes

```{r}
list_mat1 <- list("GSEA"=mat_GSEA_pvalNegLog10, 
                  "wilcoxon"=mat_wilcoxon_pvalNegLog10) 

for (name in names(list_mat1)) {
  mat <- list_mat1[[name]]
  df <- data.frame(mat)
  df <- cbind("geneset" = rownames(df), df)
  df.m <- melt(t(mat), id.vars=geneset)
  colnames(df.m) <- c("celltype","geneset","minLog10pval")
  
 p <- ggplot(df.m, aes(geneset, celltype)) + 
     geom_tile(aes(fill = minLog10pval),colour = "white") + 
     scale_fill_gradientn(colours=c("cadetblue1","red"))+
     theme(axis.text.x = element_text(angle=90))
 
 #p <- ggplotly(p)
 
  ggsave(plot=p, filename = paste0(dirPlots, prefixOut, "_",name, "_ALL_SEM_heatmap.pdf"), width=3.5, height=32)
 
}
```

of enrichment results: make boxplots

```{r}
df_mapping = load_obj("/projects/timshel/sc-genetics/sc-genetics/data/gene_annotations/gene_annotation.hsapiens_mmusculus_unique_orthologs.GRCh37.ens_v91.txt.gz")
df_NWA[["ensembl_HS"]] <- gene_map(dataIn = df_NWA[["ensembl"]], df_mapping = df_mapping, from="mmusculus_homolog_ensembl_gene", to="ensembl_gene_id", na.rm = F, replace=F)
```

```{r}
modules = c("lavenderblush", "lightpink3")
df_data <- df_SEM[,c(1, grep("^MEGLU10$|^MEGLU11$|^TEGLU17$|^DEINH3$|^MEGLU1$|^MEINH2$|^TEGLU23$|^DEGLU5$|^TEGLU4|^DEGLU4$|^TEINH12$", colnames(df_SEM)))]
df_data[["lavenderblush"]] = df_SEM[["gene"]]  %in% df_NWA[["ensembl_HS"]][df_NWA[["module"]]=="lavenderblush"]
df_data[["lightpink3"]] = df_SEM[["gene"]]  %in% df_NWA[["ensembl_HS"]][df_NWA[["module"]]=="lightpink3"]
df_data_long <- reshape::melt(df_data, id.vars=c("gene","lavenderblush", "lightpink3"))
colnames(df_data_long)[4] <- "celltype"
df_data_long
```

```{r, fig.width = 15, fig.height=10}
ggplot(data = df_data_long, aes(lavenderblush, fill=celltype, value)) +
geom_boxplot(stat = "boxplot") +
facet_wrap(facets = vars(celltype), nrow=1, scales="fixed") +
theme_minimal()
```

#### Module network plots
Use code from Gandal et al
https://github.com/mgandal/Shared-molecular-neuropathology-across-major-psychiatric-disorders-parallels-polygenic-overlap/blob/master/code/03_networkMegaAnalysis/3f_network_visualization.R


```{r}
modules = c("lavenderblush", "lightpink3")
```

```{r}
for (module in modules) {
  
  celltype = names(list_mods_uniq)[sapply(list_mods_uniq, 
                                             function(mods_uniq){
                                               module%in%mods_uniq
                                             })]
  
  
  moduleColors = list_colors_meta[[celltype]]
  datExpr = list_datExpr_meta[[celltype]]
  MEs = list_MEs_meta[[celltype]]

  modColors = data.frame(color= moduleColors,row.names = colnames(datExpr))
  cons_kme = signedKME(datExpr, MEs, outputColumnName="")
  colnames(cons_kme)=colnames(MEs) # for som reason signedKME cuts first part 
  maxsize = min(40, sum(modColors==module)) #plot top n hub genes in each module
  
  gene_idx = order(cons_kme[which(moduleColors==module),1], decreasing = T)[1:maxsize]

  gene_idx = c(gene_idx, order(cons_kme[,module], decreasing = T)[1:maxsize])
  
  hubGenes = character()
  hubGenes.kme = numeric()
  
  modgenes = colnames(datExpr)[which((modColors) == module)]
  kmes = cons_kme[modgenes, module]
  top_hubs = modgenes[order(kmes, decreasing=T)[1:maxsize]]
  top_hubs.kme = kmes[order(kmes, decreasing=T)[1:maxsize]]
  hubGenes = c(hubGenes,top_hubs)
  hubGenes.kme = c(hubGenes.kme, top_hubs.kme)

  gene_idx = match(hubGenes,colnames(datExpr))
  keepgenes = rownames(cons_kme)[gene_idx]
  
  if (FALSE) {
    adjMat = cor(datExpr)
    adjMat = adjMat[gene_idx,gene_idx]
    #topcors=0.65^9
  
    adjMat[adjMat< 0]=0
    #geneSymbols = datProbes$external_gene_id[match(keepgenes, datProbes$ensembl_gene_id)]
    geneSymbols = df_NWA[["hgnc"]][match(keepgenes,df_NWA[["ensembl"]])]
    g1 <- graph.adjacency(as.matrix(adjMat),mode="undirected",weighted=T,diag=FALSE)
    # mds = cmdscale(dist(t(adjMat)), eig = T)
    # layoutFR = mds$points
    edgecolors = numbers2colors(E(g1)$weight, colors = redWhiteGreen(100, gamma=4), signed=T, centered=T, lim=c(-1,1))
    
    
    pdf(paste0(dirPlots, prefixOut, "_", module, "_networkplot.pdf"),useDingbats = F, width=12,height=12)
    
    plot.igraph(g1, 
                vertex.label = geneSymbols,
                vertex.label.dist=0, 
                edge.width=0.25,
                vertex.size=hubGenes.kme*30, 
                vertex.frame.color="black",
                vertex.label.color="black",
                vertex.color = "white",#modColors[[1]][gene_idx],
                vertex.label.cex=1,
                layout=layout.fruchterman.reingold(g1),
                edge.color="green")
    dev.off()
  }
  
  #### NEW
  datExprSub <- datExpr[,gene_idx] 
  colnames(datExprSub) <- df_NWA[["hgnc"]][match(colnames(datExprSub),df_NWA[["ensembl"]])]
  df.corrr <- datExprSub %>% correlate(use = "pairwise.complete.obs", method="pearson")

  df.corrr %>% shave() %>% stretch(na.rm=T) -> df.corrr.long
  
  gr <- tbl_graph(nodes=data.frame(genes = unique(df.corrr.long$x)),#df.module_metadata, 
                  edges=df.corrr.long %>% filter(r>=0), directed=FALSE)
  

}
```

tidy up and package the above expr as a network plot function

```{r}

require(WGCNA)
```

```{r}

module = "lavenderblush"
celltype = names(list_mods_uniq)[sapply(list_mods_uniq,
                                           function(mods_uniq){
                                             module%in%mods_uniq
                                           })]
#celltypes=c("DEGLU4",  "DEGLU5",  "DEINH3",  "MEGLU1",  "MEGLU10", "MEGLU11", "MEINH2",  "TEGLU4",  "TEGLU17", "TEGLU23", "TEINH12")
moduleColors = list_colors_meta[[celltype]]
datExpr = list_datExpr_meta[[celltype]]
MEs = list_MEs_meta[[celltype]]
#modColors = data.frame(color= moduleColors,row.names = colnames(datExpr))
kMEs = signedKME(datExpr, MEs, outputColumnName="")
colnames(kMEs)=colnames(MEs) # for som reason signedKME cuts first part 
maxsize = min(40, sum(moduleColors==module)) #plot top n hub genes in each module
  

```

```{r}
df_mapping <- load_obj("/projects/jonatan/tools/gene_set_enrichment/mapping_mmusculus.csv")
colnames(datExpr) <- gene_map(dataIn = colnames(datExpr), df_mapping = df_mapping, from="ensembl",to="symbol", na.rm = F) 
names(moduleColors) <- gene_map(dataIn = names(moduleColors), df_mapping = df_mapping, from="ensembl",to="symbol", na.rm = F) 

```


```{r, fig.width=12,fig.height=12}
require(igraph)

plotnetwork <- function(module, moduleColors, datExpr, kMEs, maxsize=40) {
  #' @usage 
  #' @argument module 
  #' @argument modColors
  #' @argument datExpr 
  #' @argument MEs 
  #' @argument maxsize
  #' @value 

  # Get the idx and kME values of top module genes ordered by decreasing kME
  gene_idx <- order(kMEs[,module], decreasing = T)[1:maxsize]
  names(gene_idx) <- names(moduleColors)[gene_idx]
  mod_kMEs = kMEs[gene_idx, module]

  # make non-negative correlation matrix -> adjacency matrix
  adjMat = cor(datExpr[,gene_idx])
  adjMat[adjMat<0] <- 0

  # make graph
  g1 <- graph.adjacency(as.matrix(adjMat),
                        mode="undirected",
                        weighted=T,
                        diag=FALSE)
  edgecolors = numbers2colors(E(g1)$weight, colors = redWhiteGreen(100, gamma=4), signed=T, centered=T, lim=c(-1,1))
  
  #plot
  pdf(paste0(dirPlots, prefixOut, "_", module, "_networkplot.pdf"),useDingbats = F, width=12,height=12)
  plot.igraph(g1, 
              vertex.label = names(gene_idx),
              vertex.label.dist=0, 
              edge.width=0.25,
              vertex.size=mod_kMEs*30, 
              vertex.frame.color="black",
              vertex.label.color="black",
              vertex.color = "white",#modColors[[1]][gene_idx],
              vertex.label.cex=1,
              layout=layout.fruchterman.reingold(g1),
              edge.color="green")
  dev.off()
}
  
```

```{r}
df.corrr.long <- df.corrr %>%
  shave() %>% # # Convert the upper or lower triangle of a correlation data frame (cor_df) to missing values.
  stretch(na.rm = TRUE) # convert to long format. na.rm = TRUE: matrix diagonal (self-correlation) have NA values and will be dropped.
```


```{r}
for (pkg in c("gplots","kernlab","mclust","modeltools","prabclus","purrr","robustbase","scales","stringi","stringr","tidyr","trimcluster")) install.packages(pkg)
```

